### PR TITLE
[backend][sims#-1550] - Move Modules to Shared Library Part 1 of 3#1550

### PR DESCRIPTION
**Pull request by ann-aot**
- created new lib called `[@sims/integration](https://github.com/sims/integration)'
- moved IER and its services and utils to the new library
- holding `scheduler` for now, will proceed once the queue is merged

### CODE REVIEW

File: sources/packages/backend/nest-cli.json [hunk 1]
```diff
@@ -88,6 +88,15 @@
       "compilerOptions": {
         "tsConfigPath": "libs/test-utils/tsconfig.lib.json"
       }
+    },
+    "integration": {
```
Comment (andrepestana-aot) [inline sources/packages/backend/nest-cli.json:92]: "integration" is very generic. Wouldn't "fileIntegration" be better?

File: sources/packages/backend/libs/integration/src/utilities/string-builder.ts [hunk 2]
```diff
@@ -37,7 +37,7 @@ export class StringBuilder {
    */
   public appendWithEndFiller(s: string, length: number, filler: string) {
     if (s.length > length) {
-      s = s.substr(0, length);
+      s = s.substring(0, length);
```
Comment (andrepestana-aot) [inline sources/packages/backend/libs/integration/src/utilities/string-builder.ts:40]: 👍

**Review by andrepestana-aot — APPROVED** [review]
Looks good. Just suggested to change the new lib to "fileIntegration" or something more specific. 👍

File: sources/packages/backend/nest-cli.json (hunk 1 above)
Comment (ann-aot) [reply sources/packages/backend/nest-cli.json:92]: mm...this was what we discussed in gromming 
image
. devs thoughts?

File: sources/packages/backend/apps/api/src/app.system-access.module.ts [hunk 3]
```diff
@@ -65,11 +64,10 @@ import { DisbursementReceiptRequestService } from "./esdc-integration/disburseme
 import { FedRestrictionProcessingService } from "./esdc-integration/fed-restriction-integration/fed-restriction-processing.service";
 import { FedRestrictionIntegrationService } from "./esdc-integration/fed-restriction-integration/fed-restriction-integration.service";
 import { WorkflowClientService, SequenceControlService } from "@sims/services";
-import { IER12FileService } from "./institution-integration/ier-integration/ier12-file.service";
-import { IER12IntegrationService } from "./institution-integration/ier-integration/ier12-integration.service";
-
+import { IntegrationModule } from "@sims/integration";
+import { SshService } from "@sims/integration/services";
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/app.system-access.module.ts:68]: Can we add a to-do here to remove the `SshService` once all integrations are moved?

File: sources/packages/backend/apps/api/src/cra-integration/cra-files/cra-file-footer.ts [hunk 4]
```diff
@@ -1,11 +1,7 @@
...
-import {
-  DATE_FORMAT,
-  SPACE_FILLER,
-  NUMBER_FILLER,
-  TransactionCodes,
-} from "../cra-integration.models";
-import { FixedFormatFileLine } from "../../services/ssh/sftp-integration-base.models";
+import { DATE_FORMAT, NUMBER_FILLER, SPACE_FILLER } from "@sims/integration";
+import { FixedFormatFileLine } from "@sims/integration/services/ssh/sftp-integration-base.models";
+import { StringBuilder } from "@sims/integration/utilities";
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/cra-integration/cra-files/cra-file-footer.ts:3]: I would say that `StringBuilder` has nothing integration-related and I guess that it is generic enough to be moved to the generic utility library, IMO.

Comment (ann-aot) [reply sources/packages/backend/apps/api/src/cra-integration/cra-files/cra-file-footer.ts:3]: but it is only used for file integration, right?

File: sources/packages/backend/apps/api/src/cra-integration/cra-integration.models.ts [hunk 5]
```diff
@@ -1,10 +1,6 @@
 import { CRAResponseStatusRecord } from "./cra-files/cra-response-status-record";
 import { CRAResponseTotalIncomeRecord } from "./cra-files/cra-response-total-income-record";
 
-export const DATE_FORMAT = "YYYYMMDD";
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/cra-integration/cra-integration.models.ts:4]: Even though these constants are the same they were defined at the specific integration level for the purpose to allow the different integration to be changed without possibly affecting the others. Are we changing this behavior in any way here?

File: sources/packages/backend/apps/api/src/esdc-integration/disbursement-receipt-integration/disbursement-receipt-files/disbursement-receipt-file-record.ts [hunk 6]
```diff
@@ -1,5 +1,5 @@
 import { getDateOnlyFromFormat } from "@sims/utilities";
-import { DATE_FORMAT } from "../../models/esdc-integration.model";
+import { DATE_FORMAT } from "@sims/integration";
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/esdc-integration/disbursement-receipt-integration/disbursement-receipt-files/disbursement-receipt-file-record.ts:2]: These date formats are supposed to be integration specific like ESDC, CRA, and Institution. They can evolve differently and must be kept separately.

File: sources/packages/backend/apps/api/src/cra-integration/cra-integration.models.ts (hunk 5 above)
Comment (ann-aot) [reply sources/packages/backend/apps/api/src/cra-integration/cra-integration.models.ts:4]: oh....but in some files, it was imported from another integration folder, eg, 
image

File: sources/packages/backend/apps/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-file-footer.ts [hunk 7]
```diff
@@ -1,5 +1,5 @@
 import { Injectable } from "@nestjs/common";
-import { FixedFormatFileLine } from "../../../services/ssh/sftp-integration-base.models";
+import { FixedFormatFileLine } from "@sims/integration/services/ssh/sftp-integration-base.models";
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-file-footer.ts:2]: Can we move this to a index at `@sims/integration/services/ssh` maybe?

File: sources/packages/backend/apps/api/src/esdc-integration/e-cert-integration/e-cert-full-time-integration/e-cert-files/e-cert-file-footer.ts [hunk 8]
```diff
@@ -1,8 +1,5 @@
-import { StringBuilder } from "../../../../utilities/string-builder";
-import {
-  SPACE_FILLER,
-  NUMBER_FILLER,
-} from "../../../models/esdc-integration.model";
+import { StringBuilder } from "@sims/integration/utilities";
+import { NUMBER_FILLER, SPACE_FILLER } from "@sims/integration";
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/esdc-integration/e-cert-integration/e-cert-full-time-integration/e-cert-files/e-cert-file-footer.ts:2]: Same comment gave before for date formats, these ones also are supposed to belong to the integration scope.

File: sources/packages/backend/apps/api/src/esdc-integration/e-cert-integration/e-cert-full-time-integration/e-cert-full-time-integration.service.ts [hunk 9]
```diff
@@ -1,27 +1,27 @@
...
 import { ECertFullTimeFileHeader } from "./e-cert-files/e-cert-file-header";
 import { ECertFullTimeFileFooter } from "./e-cert-files/e-cert-file-footer";
 import { ECertFullTimeFileRecord } from "./e-cert-files/e-cert-file-record";
 import { DisbursementValueType, OfferingIntensity } from "@sims/sims-db";
 import { ECertIntegrationService } from "../e-cert-integration.service";
 import { ECertResponseRecord } from "../e-cert-files/e-cert-response-record";
 import { ConfigService } from "@sims/utilities/config";
+import { FixedFormatFileLine } from "@sims/integration/services/ssh/sftp-integration-base.models";
+import { SshService } from "@sims/integration/services";
+import { round } from "@sims/integration/utilities";
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/esdc-integration/e-cert-integration/e-cert-full-time-integration/e-cert-full-time-integration.service.ts:24]: These round methods can be moved to the generic `utilities` library also, IMO.

File: sources/packages/backend/libs/integration/src/integration.module.ts [hunk 10]
```diff
@@ -0,0 +1,7 @@
+import { Module } from "@nestjs/common";
+import { IER12IntegrationModule } from "./institution-integration/ier-integration/ier12-integration.module";
+
+@Module({
+  imports: [IER12IntegrationModule],
+})
+export class IntegrationModule {}
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integration/src/integration.module.ts:7]: I am not sure about having a module to aggregate all the integrations. I was under the impression that we would have one module per integration and that is it.

File: sources/packages/backend/apps/api/src/cra-integration/cra-integration.models.ts (hunk 5 above)
Comment (ann-aot) [reply sources/packages/backend/apps/api/src/cra-integration/cra-integration.models.ts:4]: nevermind. I got it. reverting

**Review by andrewsignori-aot — COMMENTED** [review]
Great work, please take a look at the comments.

File: sources/packages/backend/nest-cli.json (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/nest-cli.json:92]: Right now most of the integrations are based on SSH file uploads but they do not need to be.
I would say that any integration that is not a simple API call, like ATBC, can be created under this lib and that was the reason to try to keep this generic as `@sims/integration`. It is a valid point but I would keep it as stated in the ticket.

File: sources/packages/backend/apps/api/src/cra-integration/cra-files/cra-file-footer.ts (hunk 4 above)
Comment (ann-aot) [reply sources/packages/backend/apps/api/src/cra-integration/cra-files/cra-file-footer.ts]: moved

File: sources/packages/backend/apps/api/src/cra-integration/cra-integration.models.ts (hunk 5 above)
Comment (ann-aot) [reply sources/packages/backend/apps/api/src/cra-integration/cra-integration.models.ts]: updated

**Review by andrewsignori-aot — COMMENTED** [review]
Thanks for doing the changes, everything looks good but the Sonarcloud issues about the duplicated imports. Once they are resolved I can approve the PR.

**Review by andrewsignori-aot — APPROVED** [review]
Looks good 👍

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @ann-aot

**Pull request merged**